doc/notes/cvar.txt

```diff
@@ -27,7 +27,7 @@ If the type is enum, the possible values of the enum need be listed in the descr
 Above example represents (in `src/include/mpir_cvars.h`):
     extern int MPIR_CVAR_IALLGATHER_INTRA_ALGORITHM;
     enum MPIR_CVAR_IALLGATHER_INTRA_ALGORITHM_choice {
-        MPIR_CVAR_IALLGATHER_INTRA_ALGORITHM_auto,
+        MPIR_CVAR_IALLGATHER_INTRA_ALGORITHM_json,
```

Comment:
I am not sure it is a good idea to rename from `_auto` to `_json`. A function name and signature is mostly read from a caller's context. In the caller's semantics, it cares about whether it is a specific algorithm or an "auto"-selection algorithm. Now for the implementation of an "auto" algorithm, it could be hard coded, using a `json` selection, or using an "AI". But the implementation detail is irrelevant in the caller's context, right?

Reply:
That's fair.  I can rename it to `_auto`.  The fact that the "auto" logic was generated from a json file is irrelevant here.

Reply:
Hmm.  Maybe we need better naming here.  When the user wants to provide a `json` file, she would set `MPIR_CVAR_COLL_SELECTION_TUNING_JSON_FILE`.  What happens when the user picks `MPIR_CVAR_IALLGATHER_INTRA_ALGORITHM=auto` and then sets `MPIR_CVAR_COLL_SELECTION_TUNING_JSON_FILE` to some CVAR file?  Should I pick the "auto" (internal) algorithm or the json file?

One possibility is to say that setting `MPIR_CVAR_COLL_SELECTION_TUNING_JSON_FILE` will override all `MPIR_CVAR_IALLGATHER_INTRA_ALGORITHM` settings.  If we decide to do this, the CVAR descriptions need to be updated.

Reply:
I don't see a problem here. `MPIR_CVAR_IALLGATHER_INTRA_ALGORITHM=auto` chooses a specific algorithm, always takes precedence. `MPIR_CVAR_COLL_SELECTION_TUNING_JSON_FILE` is an inner-level control knob that tunes how the "auto" algorithm should be done. The latter CVAR only remain effective when the specific algorithm is "auto", default or specifically chosen.

Reply:
That's fine.  I was suggesting giving `MPIR_CVAR_COLL_SELECTION_TUNING_JSON_FILE` giving higher priority and you are suggesting giving `MPIR_CVAR_IALLGATHER_INTRA_ALGORITHM` higher priority.  Either way is fine, but we still need to edit the descriptions to mention that.  I can take care of it.  But in that case, I'd recommend that we use `_json` instead of `_auto`, because when the user specifies that, it's no longer automatic but is specified by the user.

Reply:
> But in that case, I'd recommend that we use `_json` instead of `_auto`, because when the user specifies that, it's no longer automatic but is specified by the user.

What if we make `_json` a static internal function and have `_auto` call `_json`. Theoretically, `_auto` may call some other "auto" method in stead.

It seems I am still missing some of your reasoning. In my mind, `MPIR_CVAR_COLL_SELECTION_TUNING_JSON_FILE` is an option controls the `_json` method. Whether `_json` method is being used is outside that CVAR.

Reply:
> Theoretically, _auto may call some other "auto" method in stead.

That's exactly my concern.  Since `_auto` can call something else instead of `_json`, the user cannot force it to use `_json`.

Reply:
Now I see. In that case, we'll need a separate CVAR, e.g. `MPIR_CVAR_COLL_USE_JSON`